src/Sentry/Protocol/Transaction.cs

```diff
@@ -0,0 +1,179 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.Json;
+using Sentry.Extensibility;
+using Sentry.Internal.Extensions;
+
+namespace Sentry.Protocol
+{
+    // https://develop.sentry.dev/sdk/event-payloads/transaction
+    public class Transaction : ISpan, IJsonSerializable
+    {
+        private readonly IHub _hub;
+
+        public string Name { get; }
+        public SentryId SpanId { get; }
+        public SentryId? ParentSpanId { get; }
+        public SentryId TraceId { get; set; }
+        public DateTimeOffset StartTimestamp { get; set; }
+        public DateTimeOffset EndTimestamp { get; set; }
+        public string Operation { get; }
+        public string? Description { get; set; }
+        public SpanStatus? Status { get; set; }
+        public bool IsSampled { get; set; }
+
+        private Dictionary<string, string>? _tags;
+        public IReadOnlyDictionary<string, string> Tags => _tags ??= new Dictionary<string, string>();
+
+        private Dictionary<string, object>? _data;
+        public IReadOnlyDictionary<string, object> Data => _data ??= new Dictionary<string, object>();
+
+        private List<Span>? _children;
+        public IReadOnlyList<Span> Children => _children ??= new List<Span>();
+
+        internal Transaction(
+            IHub hub,
+            string name,
+            SentryId? spanId = null,
+            SentryId? parentSpanId = null,
+            string operation = "unknown")
+        {
+            _hub = hub;
+            Name = name;
+            SpanId = spanId ?? SentryId.Create();
+            ParentSpanId = parentSpanId;
+            TraceId = SentryId.Create();
+            StartTimestamp = EndTimestamp = DateTimeOffset.Now;
+            Operation = operation;
+        }
+
+        public Transaction(IHub hub, string name, string operation)
+            : this(
+                hub,
+                name,
+                null,
+                null,
+                operation)
+        {
+        }
+
+        public ISpan StartChild(string operation)
+        {
+            var span = new Span(null, SpanId, operation);
+            (_children ??= new List<Span>()).Add(span);
```

Comment:
This needs to be synchronized, if we're allowing the transaction instance to be passed to multiple threads where we'll be calling `StartChild`

Reply:
storing Spans on AsyncLocal works great for this case.

Reply:
> This needs to be synchronized...

Do you mean synchronizing the list creation specifically, or turning the list into a `ConcurrentBag`?

Reply:
two concurrent calls to `StartChild` would yield multiple `List<Span>`'s created, and due to CPU cache and other optimizations it's not guaranteed that the instance set first would be visible to all threads. This particular case could be solved with `volatile` but that wouldn't solve everything:

We'd have a non thread-safe list in the field, which would throw as we read spans in the scope and concurrently try to add to it (call StartSpan) again.

Without much looking into it the simplest way I can think is to indeed use `ConcurrentBag` (if order doesn't matter), but have that initialized in the constructor (or with a `Lazy<T>` with `PublicationOnly`). 

If you prefer we can hop on a call and brainstorm this together. I'm not familiar with the performance API to the level I can just drop a comment here and be confident we're addressing all the issues. /cc @marandaneto FYI